File: solutions/go/1143_longest_common_subsequence.go
package main

// Problem: 1143 Longest Common Subsequence
// URL: https://leetcode.cn/problems/longest-common-subsequence/
// Difficulty: Medium
// Tags: String, Dynamic Programming
// Constraints:
// - 1 <= text1.length, text2.length <= 1000
// - text1 and text2 consist of only lowercase English characters.

import "reflect"

// === Solution ===
func longestCommonSubsequence(text1 string, text2 string) int {
	m, n := len(text1), len(text2)
	dp := make([][]int, m+1)
	for i, _ := range dp {
		dp[i] = make([]int, n+1)
	}

	for i, _ := range text1 {
		for j, _ := range text2 {
			if text1[i] == text2[j] {
				dp[i+1][j+1] = dp[i][j] + 1
			} else {
				dp[i+1][j+1] = max(dp[i+1][j], dp[i][j+1])
			}
		}
	}

	return dp[m][n]
}

// === Test Code ===
// Example 1
// Input: "abcde"; "ace"
// Output: 3
// Example 2
// Input: "abc"; "abc"
// Output: 3
// Example 3
// Input: "abc"; "def"
// Output: 0

func selfTest() {
	if got := longestCommonSubsequence("abcde", "ace"); !reflect.DeepEqual(got, 3) {
		panic("example 1 failed")
	}
	if got := longestCommonSubsequence("abc", "abc"); !reflect.DeepEqual(got, 3) {
		panic("example 2 failed")
	}
	if got := longestCommonSubsequence("abc", "def"); !reflect.DeepEqual(got, 0) {
		panic("example 3 failed")
	}
}

func selfTestAdditional() {
	// 最小长度边界：单字符且相等 -> 1
	if got := longestCommonSubsequence("a", "a"); !reflect.DeepEqual(got, 1) {
		panic("single equal char failed")
	}
	// 单字符不相等 -> 0
	if got := longestCommonSubsequence("a", "b"); !reflect.DeepEqual(got, 0) {
		panic("single distinct char failed")
	}
	// 顺序敏感：公共字符集是 {a,c,e} 但顺序不允许全部保留，LCS 是 "ae"/"ac" -> 2
	if got := longestCommonSubsequence("abcde", "aec"); !reflect.DeepEqual(got, 2) {
		panic("order-sensitive case failed")
	}
	// 重复字符：结果受较短串的字符数量上限约束 -> 2
	if got := longestCommonSubsequence("aaaa", "aa"); !reflect.DeepEqual(got, 2) {
		panic("repeated char case failed")
	}
	// 完全反序：只能取到单个字符 -> 1
	if got := longestCommonSubsequence("abc", "cba"); !reflect.DeepEqual(got, 1) {
		panic("reversed case failed")
	}
}

func main() {
	selfTest()
	selfTestAdditional()
}

File: solutions/py/1143_longest_common_subsequence.py
from __future__ import annotations

# Problem: 1143 Longest Common Subsequence
# URL: https://leetcode.cn/problems/longest-common-subsequence/
# Difficulty: Medium
# Tags: String, Dynamic Programming
# Constraints:
# - 1 <= text1.length, text2.length <= 1000
# - text1 and text2 consist of only lowercase English characters.


# === Solution ===
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        m, n = len(text1), len(text2)
        dp = [[0] * (n + 1) for _ in range(m + 1)]

        for i1, c1 in enumerate(text1):
            for i2, c2 in enumerate(text2):
                if c1 == c2:
                    dp[i1 + 1][i2 + 1] = dp[i1][i2] + 1
                else:
                    dp[i1 + 1][i2 + 1] = max(dp[i1][i2 + 1], dp[i1 + 1][i2])

        return dp[m][n]


# === Test Code ===
# Example 1
# Input: "abcde"; "ace"
# Output: 3
# Example 2
# Input: "abc"; "abc"
# Output: 3
# Example 3
# Input: "abc"; "def"
# Output: 0


def _run_examples() -> None:
    solver = Solution()
    assert solver.longestCommonSubsequence("abcde", "ace") == 3, "example 1 failed"
    assert solver.longestCommonSubsequence("abc", "abc") == 3, "example 2 failed"
    assert solver.longestCommonSubsequence("abc", "def") == 0, "example 3 failed"


def _run_additional_tests() -> None:
    solver = Solution()
    # 最小长度边界：单字符且相等 -> 1
    assert solver.longestCommonSubsequence("a", "a") == 1
    # 单字符不相等 -> 0
    assert solver.longestCommonSubsequence("a", "b") == 0
    # 顺序敏感：公共字符集是 {a,c,e} 但顺序不允许全部保留，LCS 是 "ae"/"ac" -> 2
    assert solver.longestCommonSubsequence("abcde", "aec") == 2
    # 重复字符：结果受较短串的字符数量上限约束 -> 2
    assert solver.longestCommonSubsequence("aaaa", "aa") == 2
    # 完全反序：只能取到单个字符 -> 1
    assert solver.longestCommonSubsequence("abc", "cba") == 1


if __name__ == "__main__":
    _run_examples()
    _run_additional_tests()
